docs/source/perf/spmd_basic.md

```diff
@@ -61,56 +65,133 @@ SPMD with other distributed libraries.
 
 ### Mesh
 
-For a given cluster of devices, a physical mesh is a representation of
-the interconnect topology.
+The SPMD programming model is built around the concept of a device mesh, commonly
+referred to as a mesh. A device mesh is a logical N-dimensional arrangement of
+compute devices (e.g. TPU cores) where the axes are used as shorthands to
+define sharded computation. It is logical in the sense that the mesh shape does
+not necessarily have to reflect the physical network layout. You could create
+different kinds of device meshes over the same set of physical devices. For
+example, a 512-core TPU slice could be treated as a 3D mesh of 16×16×2, a 2D mesh
+of 32×16, or a 1D mesh of 512, depending on how we want to partition tensors. You
+may use the `Mesh` class to create a device mesh.
+
+In the following snippet:
+
+```python
+mesh_shape = (num_devices, 1)
+device_ids = np.array(range(num_devices))
+mesh = Mesh(device_ids, mesh_shape, ('data', 'model'))
+```
 
-1.  `mesh_shape` is a tuple that will be multiplied to the total number
-    of physical devices.
-2.  `device_ids` is almost always `np.array(range(num_devices))`.
-3.  Users are also encouraged to give each mesh dimension a name. In the
-    above example, the first mesh dimension is the `data` dimension and
-    the second mesh dimension is the `model` dimension.
+- `mesh_shape` is a tuple whose elements must multiply to the total number
+  of physical devices in the environment.
+- `device_ids` specifies the ordering of physical devices within the mesh in row
+  major order. It is always a flat numpy array of integers ranging from `0` to
+  `num_devices - 1`, where `num_devices` is the total number of devices in the
+  environment. Each ID indexes into the list of devices in
+  `xr.global_runtime_device_attributes()`.
+  The simplest ordering is `np.array(range(num_devices))`, but tuning the
+  device ordering and the mesh shape to utilize the underlying physical
+  interconnect can improve efficiency.
+- Users are also encouraged to give each mesh axis a name. Later we can
```

Comment:
Best practice is to give each mesh axis a name. You can then shard each dimension of tensors over specific mesh axes ....

Reply:
done